crates/prune/src/pruner.rs

```diff
@@ -196,133 +314,141 @@ impl<DB: Database> Pruner<DB> {
             PrunePart::Receipts,
-            PruneCheckpoint { block_number: to_block, prune_mode },
+            PruneCheckpoint {
+                block_number: last_pruned_block,
+                tx_number: Some(last_pruned_transaction),
+                prune_mode,
+            },
         )?;
 
-        Ok(())
+        Ok(done)
     }
 
-    /// Prune transaction lookup entries up to the provided block, inclusive.
+    /// Prune transaction lookup entries up to the provided block, inclusive, respecting the batch
+    /// size.
     #[instrument(level = "trace", skip(self, provider), target = "pruner")]
     fn prune_transaction_lookup(
         &self,
         provider: &DatabaseProviderRW<'_, DB>,
         to_block: BlockNumber,
         prune_mode: PruneMode,
     ) -> PrunerResult {
-        let range = match self.get_next_tx_num_range_from_checkpoint(
+        let (start, end) = match self.get_next_tx_num_range_from_checkpoint(
             provider,
             PrunePart::TransactionLookup,
             to_block,
         )? {
             Some(range) => range,
             None => {
                 trace!(target: "pruner", "No transaction lookup entries to prune");
-                return Ok(())
-            }
-        };
-        let last_tx_num = *range.end();
-        let total = range.clone().count();
-        let mut processed = 0;
-
-        for i in range.step_by(self.batch_sizes.transaction_lookup) {
-            // The `min` ensures that the transaction range doesn't exceed the last transaction
-            // number. `last_tx_num + 1` is used to include the last transaction in the range.
-            let tx_range = i..(i + self.batch_sizes.transaction_lookup as u64).min(last_tx_num + 1);
-
-            // Retrieve transactions in the range and calculate their hashes in parallel
-            let mut hashes = provider
-                .transactions_by_tx_range(tx_range.clone())?
-                .into_par_iter()
-                .map(|transaction| transaction.hash())
-                .collect::<Vec<_>>();
-
-            // Number of transactions retrieved from the database should match the tx range count
-            let tx_count = tx_range.clone().count();
-            if hashes.len() != tx_count {
-                return Err(PrunerError::InconsistentData(
-                    "Unexpected number of transaction hashes retrieved by transaction number range",
-                ))
+                return Ok(true)
             }
+        }
+        .into_inner();
+        let tx_range = start..=(end.min(start + self.batch_sizes.transaction_lookup as u64 - 1));
+        let tx_range_end = *tx_range.end();
+
+        // Retrieve transactions in the range and calculate their hashes in parallel
+        let hashes = provider
+            .transactions_by_tx_range(tx_range.clone())?
+            .into_par_iter()
+            .map(|transaction| transaction.hash())
+            .collect::<Vec<_>>();
+
```

Comment:
can we add back the sorting? or did you test and had little difference?

Reply:
we're saving the `last_pruned_transaction`, and if we sort the hashes, it will stop being the `last_pruned_transaction` according to the initial order of `tx_range`